### DISCUSSION TITLE
feat: pass reason for closing to the emitted event

### DISCUSSION THREAD

**Discussion by JohnCampionJr** [tailwindlabs/headlessui#2812] [Ideas]
Can we please get this?  It's a pretty common use case at this point to not allow clicking outside the dialog to close it.  I think @mgambill's suggestion is the best way to implement it with non-breaking API and no extra properties or anything needed.

What if we just pass information as part of the emit event then the user will have information to determine if they want to close or not. @adamwathan 

`api.close( reason ) // 'backdrop' | 'key' | 'trap' | 'other')`

```js
function onClose(e) {
   if (e.reason === 'backdrop') // update ref or not 
}
```

_Originally posted by @mgambill in https://github.com/tailwindlabs/headlessui/issues/1968#issuecomment-1722745220_

**Comment by RobinMalfait** [comment]
Hey!

Thanks for this suggestion. Right now passing in a reason would result in a breaking change for components such as the `Dialog` component. This is because the `onClose` is called with `false`. This is done so that you can use this short convenient way of closing a dialog:

```js
let [isOpen, setIsOpen] = useState(false)
<Dialog open={isOpen} onClose={setIsOpen} />
```

We've talked about this exact feature internally. One way of solving this is by introducing another method like `onDismiss` that does get more information about _why_ something is dismissed.

Note: you have to be careful from an accessibility perspective that you don't change the behaviour without very good reason. For example, a sighted user can click outside, somebody using Assistive Technology will likely press <kbd>esc</kbd> instead. Changing the behaviour between those interactions will degrade user experience.

That said, I believe @thecrypticace is potentially going to look into this this week.

**Reply by JohnCampionJr** [reply]
I understand the ESC situation.  It's more about the accidental clicks outside of an important dialog.  It's much harder to accidentally hit ESC.

I appreciate any solution you can provide 😀

**Reply by TomCaserta** [reply]
Is this still under consideration? 

I have a similar problem @JohnCampionJr mentions, where if you have form elements inside a dialog its very easy to accidentally close the modal. Especially when you have a select element inside the modal, often people will click outside of the select to close it without thinking and doing so would erase the entire form when it closes the modal as well.

**Comment by kulapoo** [comment]
yo @JohnCampionJr ,

Im not sure if this will work for u,  please check if u find any cons. That will help me as well or issue that I might encounter in doing this

what I did is something like this


```js
import { Dialog as Dialog Base } from ''

// using styledComponent in my case or can use any class manipulation
const DialogRoot = styled(DialogBase)`
  z-index: 99999;
  position: relative
  display: ${props => props.isOpen ? 'block' : 'none'};

`
function CustomDialog(props) {
   const {
     open,
     onClose,
     ...otherProps
   } = props
   
   const closeWorkAround = useCallback(() => {}, []) // do nothing
   return (
     <DialogRoot 
       static // this is will do the trick
       isOpen={open}
       open={open}
       onClose={closeWorkAround}
     >
       
        // rest of the requiredComponents
     </DialogRoot>
   )
}
`
```

[sample codesandbox](https://codesandbox.io/s/immutable-pond-j96m26?file=/src/App.js)
